#70 Implement logout endpoint

devdsys opened the pull request:
Add logout endpoint to blacklist refresh token and shorten token lifespan

Link to Issue: #70

VovaD199 commented on forum/users/views.py:
```diff
@@ -1,3 +1,51 @@
+                return Response({"error": "Access denied."}, status=403)
+            else:
+                return Response({"error": "Invalid or expired token."}, status=403)
+               
```
The current implementation compares the exception text (str(e)), which is not reliable.
It is better to check the exception type:
from rest_framework_simplejwt.exceptions import TokenError

except TokenError as e:
    logger.error(f"JWT Token error occurred: {e}")
    return Response({"error": str(e)}, status=403)

devdsys replied:
I have updated the exception handling logic to avoid exposing business logic or error details to clients. Now, in case of validation or other errors, the client receives an 'Access Denied' response. However, all errors are logged for developers.

VovaD199 commented on forum/users/views.py:
```diff
@@ -1,3 +1,51 @@
+        try:
+            token = RefreshToken(refresh_token)
+            token.blacklist()
+
```
Add logging of successful logout for easy monitoring:
logger.info("User logged out successfully.")

devdsys replied:
Done.

VovaD199 commented on forum/users/views.py:
```diff
@@ -1,3 +1,51 @@
+
+class LogoutView(APIView):
+    valid_exceptions = ['Token is blacklisted', 'Token is invalid or expired', 'Token has wrong type']
+
```
The name of the variable valid_exceptions:

The name can be misleading, as these are text messages, not exceptions themselves. You might want to change it to, for example, valid_exception_messages.

devdsys replied:
I have updated the exception handling logic, so this line is now removed.

devdsys commented:
I have added the IsAuthenticated permission to the view. Additionally, I have included refresh and access token validations in the serializer, as well as a comparison of the token's 'user_id'. I have also updated the exception handling logic to prevent exposing business logic or error details to clients. In case of validation or other errors, the client now receives an 'Access Denied' response, while all errors are logged for developers.

IraMladanovych02 reviewed (approved):
Good job!
